### PR TITLE
Add utils for domain backup function

### PR DESCRIPTION
**Pull request by lento-sun**
Libvirt's incremental backup related supporting methods added in this file.

Signed-off-by: Yi Sun <yisun@redhat.com>

### CODE REVIEW

**Comment by dzhengfy** [comment]
@lento-sun , please help solve the travis failure.

**Comment by lento-sun** [comment]
Depends on: https://github.com/avocado-framework/avocado-vt/pull/2503

**Comment by lento-sun** [comment]
> @lento-sun , please help solve the travis failure.

This depends on: https://github.com/avocado-framework/avocado-vt/pull/2503 which is not merged yet, so has "E0611:  9,0: : No name 'backup_xml' in module 'virttest.libvirt_xml'"

File: virttest/utils_backup.py [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+
+    :param cp_params: Params of checkpoint
+    :param disk_param_list: A list of disk params used in checkpoint xml
+    :return: The checkponit xml
+    """
+    cp_xml = CheckpointXML()
+    cp_name = cp_params.get("checkpoint_name")
+    cp_desc = cp_params.get("checkpoint_desc")
+    if not cp_name:
+        logging.error("Checkpoint name must be provided")
```
Comment (chunfuwen) [inline virttest/utils_backup.py:49]: Better throw exception directly

File: virttest/utils_backup.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        cp_xml.description = cp_desc
+    if disk_param_list:
+        if type(disk_param_list) is not list:
+            logging.error("checkpoint disk tags should be defined by a "
+                          "list.")
+            return
+        cp_xml.disks = disk_param_list
+    cp_xml.xmltreefile.write()
+
+    def file_exists():
```
Comment (chunfuwen) [inline virttest/utils_backup.py:62]: file_exists() to _file_exists() ?

File: virttest/utils_backup.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    cp_name = cp_params.get("checkpoint_name")
+    cp_desc = cp_params.get("checkpoint_desc")
+    if not cp_name:
+        logging.error("Checkpoint name must be provided")
+        return
+    cp_xml.name = cp_name
+    if cp_desc:
+        cp_xml.description = cp_desc
+    if disk_param_list:
+        if type(disk_param_list) is not list:
```
Comment (chunfuwen) [inline virttest/utils_backup.py:55]: isinstance( ) is more commonly used.

File: virttest/utils_backup.py [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    backup_xml = BackupXML()
+    backup_mode = backup_params.get("backup_mode")
+    backup_incremental = backup_params.get("backup_incremental")
+    backup_server = backup_params.get("backup_server")
+    if backup_mode:
+        backup_xml.mode = backup_mode
+    if backup_incremental:
+        backup_xml.incremental = backup_incremental
+    if backup_server:
+        if type(backup_server) is not dict:
```
Comment (chunfuwen) [inline virttest/utils_backup.py:86]: isinstance(xx,xx) can be more common used

File: virttest/utils_backup.py [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+            logging.error("disk driver tag should be defined by a dict.")
+            return
+        backup_disk_xml.driver = backup_driver
+    if backup_scratch:
+        if type(backup_scratch) is not dict:
+            logging.error("disk scratch tag should be defined by a dict.")
+        backup_disk_xml.scratch = backup_scratch
+    backup_disk_xml.xmltreefile.write()
+
+    def file_exists():
```
Comment (chunfuwen) [inline virttest/utils_backup.py:144]: there are 3 places to define file_exists(), maybe pull out as common function

File: virttest/utils_backup.py [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+        process.run(cmd, shell=True)
+
+
+def cmp_backup_data(original_file, backup_file,
+                    original_file_driver='qcow2', backup_file_driver='qcow2'):
+    """
+    Compare the backuped data and the original data
+
+    :param original_file: The file containing the original data
+    :param backup_file: The file containing the backuped data
```
Comment (chunfuwen) [inline virttest/utils_backup.py:287]: miss parameter docstring

File: virttest/utils_backup.py [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+    Dump real data from the source file to a target file according to data map
+
+    :param data_map: The data map used to indicate which part of the image
+    containing real data
+    :param source_file_path: The path to the source file
+    :param target_file_path: The path to the target file
+    """
+    with open(target_file_path, 'w'):
+        pass
+    cmd = "qemu-io -f %s -c \"read -vC %s %s\" %s -r -U  | sed '$d'"
```
Comment (chunfuwen) [inline virttest/utils_backup.py:274]: should all below logic be put under with open block?

File: virttest/utils_backup.py (hunk 5 above)
Comment (lento-sun) [reply virttest/utils_backup.py]: directly use "utils_misc.wait_for(lambda: os.path.exists(cp_xml.xml), 5)" now, thx

File: virttest/utils_backup.py (hunk 7 above)
Comment (lento-sun) [reply virttest/utils_backup.py]: yes, updated, thx

**Comment by dzhengfy** [comment]
@yicui, please help review.

File: virttest/utils_backup.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    result = process.run(cmd, shell=True).stdout_text
+    json_data = json.loads(result)
+    data_map = []
+    for entry in json_data:
+        if entry["data"]:
+            data_map.append(entry)
+    return data_map
+
+
+def dd_data_to_vm_disk(vm, disk, bs, seek, count):
```
Comment (Yingshun) [inline virttest/utils_backup.py:231]: Would you please move it to other files, it's not very close to backup feature I think, how do you think?

File: virttest/utils_backup.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+    for entry in json_data:
+        if entry["data"]:
+            summary += entry["length"]
+    return summary/1024/1024
+
+
+def get_img_data_map(img_path, driver='qcow2'):
+    """
+    Get the data map of a image
+
```
Comment (Yingshun) [inline virttest/utils_backup.py:217]: Same as above

File: virttest/utils_backup.py [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    :return: Trun if data is correct, false if not
+    """
+    original_file_dump = original_file + ".dump"
+    backup_file_dump = backup_file + ".dump"
+    data_map = get_img_data_map(backup_file, backup_file_driver)
+    dump_data_to_file(data_map, original_file, original_file_dump, original_file_driver)
+    dump_data_to_file(data_map, backup_file, backup_file_dump, backup_file_driver)
+    if filecmp.cmp(original_file_dump, backup_file_dump):
+        return True
+    return False
```
Comment (Yingshun) [inline virttest/utils_backup.py:285]: It's redundant, would you please remove it?

File: virttest/utils_backup.py (hunk 8 above)
Comment (lento-sun) [reply virttest/utils_backup.py:231]: will move to utils_disk.py

File: virttest/utils_backup.py (hunk 9 above)
Comment (lento-sun) [reply virttest/utils_backup.py:217]: This method is not to get the image's entire map, but only get the list of 'data' entries (excluding 'zero' entries) from a backuped file, and then, do a data dump in cmp_backup_data() method with this list. So it's not quite useful for other features till now. We don't do such kind of bit-level data comparison for whole images in other features. So leave it here could be OK.

File: virttest/utils_backup.py (hunk 10 above)
Comment (lento-sun) [reply virttest/utils_backup.py:285]: will remove

**Review by Yingshun — APPROVED** [review]

**Review by chunfuwen — APPROVED** [review]
LGTM

**Pull request merged**